### PR TITLE
[12.0][IMP] procurement_purchase_no_grouping: Add product_category in procured_purchase_grouping field

### PR DESCRIPTION
**Pull request by victoralmau**
Add product_category in `procured_purchase_grouping` field

@Tecnativa TT27393

### CODE REVIEW

File: procurement_purchase_no_grouping/models/stock_rule.py [hunk 1]
```diff
@@ -15,16 +16,19 @@ class StockRule(models.Model):
...
-            if values.get("move_dest_ids"):
-                domain += (("id", "=", -values["move_dest_ids"][:1].id),)
-            # The minimum is imposed by PG int4 limit
-            domain += (("id", "=", random.randint(-2147483648, 0)),)
+        if not self.env.context.get('skip_domain'):
+            if self.env.context.get('grouping', 'standard') == 'order':
+                if values.get("move_dest_ids"):
+                    domain += (("id", "=", -values["move_dest_ids"][:1].id),)
+                # The minimum is imposed by PG int4 limit
+                domain += (("id", "=", random.randint(-2147483648, 0)),)
```
Comment (Tardo) [inline procurement_purchase_no_grouping/models/stock_rule.py]: What kind of magic is being done here? negative ids? O_o
```suggestion
        if not self.env.context.get('skip_domain') and self.env.context.get('grouping', 'standard') == 'order':
            if values.get("move_dest_ids"):
                domain += (("id", "=", -values["move_dest_ids"][:1].id),)
            # The minimum is imposed by PG int4 limit
            domain += (("id", "=", random.randint(-2147483648, 0)),)
```

File: procurement_purchase_no_grouping/models/product_category.py [hunk 2]
```diff
@@ -11,7 +11,8 @@ class ProductCategory(models.Model):
     procured_purchase_grouping = fields.Selection(
         [('standard', 'Standard grouping'),
          ('line', 'No line grouping'),
-         ('order', 'No order grouping')],
+         ('order', 'No order grouping'),
+         ('product_category', 'Product category')],
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/product_category.py]:
```suggestion
         ('product_category', 'Product category grouping')],
```

File: procurement_purchase_no_grouping/models/product_category.py [hunk 3]
```diff
@@ -22,4 +23,5 @@ class ProductCategory(models.Model):
              "the same supplier, it will be reused, but lines won't be "
              "merged.\n"
              "* No order grouping: This option will prevent any kind of "
-             "grouping.")
+             "grouping.\n"
+             "* Product category: This option grouping lines by product category.")
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/models/product_category.py]:
```suggestion
             "* Product category grouping: This option groups products in the same purchase order that belongs to the same product category.")
```

File: procurement_purchase_no_grouping/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -6,3 +6,4 @@ Go to each product category, and select one of these values in the field
 * *No line grouping*: With this value, if there are any open purchase order
   for the same supplier, it will be reused, but lines won't be merged.
 * *No order grouping*: This option will prevent any kind of grouping.
+* *Product category*: This option grouping lines by product category.
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/readme/CONFIGURE.rst]:
```suggestion
* *Product category grouping*: This option groups products in the same purchase order that belongs to the same product category.
```

File: procurement_purchase_no_grouping/models/stock_rule.py (hunk 1 above)
Comment (pedrobaeza) [reply procurement_purchase_no_grouping/models/stock_rule.py]: The trick is used for having no results at all from the search.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
You are checking nowhere the category of the product against the category of the existing draft purchase orders, so this can't work and the test you have introduced is not enough. You have to test both positive and negative matching to see that everything works.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Please change Spanish translation + POT for completion, and the test is still not testing the new feature. It has to test:

- Having 2 products with same vendor and same category and running the procurement for both, there's only one PO for both.
- Having 2 products with same vendor and different category, there are 2 POs.

File: procurement_purchase_no_grouping/i18n/es.po [hunk 5]
```diff
@@ -8,35 +8,41 @@ msgid ""
...
 #: selection:product.category,procured_purchase_grouping:0
 msgid "No order grouping"
-msgstr ""
+msgstr "Sin agrupación de pedidos"
 
 #. module: procurement_purchase_no_grouping
 #: model:ir.model.fields,field_description:procurement_purchase_no_grouping.field_product_category__procured_purchase_grouping
 msgid "Procured purchase grouping"
-msgstr ""
+msgstr "Agrupación por pedido de compra"
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/i18n/es.po]:
```suggestion
msgstr "Agrupación de la compra abastecida"
```

File: procurement_purchase_no_grouping/i18n/es.po [hunk 6]
```diff
@@ -52,18 +58,30 @@ msgid ""
...
 #: selection:product.category,procured_purchase_grouping:0
 msgid "Standard grouping"
-msgstr ""
+msgstr "Agrupación estándar"
 
 #. module: procurement_purchase_no_grouping
 #: model:ir.model,name:procurement_purchase_no_grouping.model_stock_rule
 msgid "Stock Rule"
-msgstr ""
+msgstr "Regla de Inventario"
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/i18n/es.po]: On Spanish there's no upper in the middle of the sentence:
```suggestion
msgstr "Regla de inventario"
```

File: procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py [hunk 7]
```diff
@@ -89,3 +90,42 @@ def test_procurement_no_grouping_order_purchase(self):
...
+        product2 = self.product.copy()
+        product2.write({
+            'seller_ids': [
+                (0, 0, {
+                    'name': self.partner.id,
+                    'min_qty': 1.0,
+                })
+            ]
+        })
+        product2.categ_id = category2.id
```
Comment (pedrobaeza) [inline procurement_purchase_no_grouping/tests/test_procurement_purchase_no_grouping.py]: Why not including this in the previous write?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Almost there!

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@Tardo please update your review

**Review by marcelsavegnago — APPROVED** [review]
LGTM

**Comment by pedrobaeza** [comment]
/ocabot merge major

Please fw-port it to v13

**Pull request merged**